genie-web/src/main/java/com/netflix/genie/web/services/impl/S3AttachmentService.java

```diff
@@ -0,0 +1,225 @@
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.services.impl;
+
+import com.amazonaws.SdkClientException;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3URI;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.netflix.genie.common.internal.aws.s3.S3ClientFactory;
+import com.netflix.genie.web.exceptions.checked.AttachmentTooLargeException;
+import com.netflix.genie.web.exceptions.checked.SaveAttachmentException;
+import com.netflix.genie.web.properties.AttachmentServiceProperties;
+import com.netflix.genie.web.services.AttachmentService;
+import com.netflix.genie.web.util.MetricsUtils;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.core.io.Resource;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Implementation of the AttachmentService interface which saves attachments to AWS S3.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class S3AttachmentService implements AttachmentService {
+
+    private static final String METRICS_PREFIX = "genie.jobs.attachments.s3";
+    private static final String COUNT_DISTRIBUTION = METRICS_PREFIX + ".count.distribution";
+    private static final String LARGEST_SIZE_DISTRIBUTION = METRICS_PREFIX + ".largest.distribution";
+    private static final String TOTAL_SIZE_DISTRIBUTION = METRICS_PREFIX + ".totalSize.distribution";
+    private static final String SAVE_TIMER = METRICS_PREFIX + ".upload.timer";
+    private static final Set<URI> EMPTY_SET = ImmutableSet.of();
+    private static final String SLASH = "/";
+    private static final String S3 = "s3";
+    private final S3ClientFactory s3ClientFactory;
+    private final AttachmentServiceProperties properties;
+    private final MeterRegistry meterRegistry;
+    private final AmazonS3URI s3BaseURI;
+
+    /**
+     * Constructor.
+     *
+     * @param s3ClientFactory             the s3 client factory
+     * @param attachmentServiceProperties the service properties
+     * @param meterRegistry               the meter registry
+     */
+    public S3AttachmentService(
+        final S3ClientFactory s3ClientFactory,
+        final AttachmentServiceProperties attachmentServiceProperties,
+        final MeterRegistry meterRegistry
+    ) {
+        this.s3ClientFactory = s3ClientFactory;
+        this.properties = attachmentServiceProperties;
+        this.meterRegistry = meterRegistry;
+        this.s3BaseURI = new AmazonS3URI(attachmentServiceProperties.getLocationPrefix());
```

Comment:
You have it as `AttachmentServiceProperties` but then make an assumption here that it's a S3 path. This will throw a runtime exception if it's not a valid S3 URI so the system will fail to start so I guess that's fine just pointing out the assumption that I guess you're saying the deployment needs to be consistent

Reply:
Personally I wouldn't call this an assumption. It's a bean validating a configuration parameter, and failing to initialize if the value is misconfigured for the bean type.